src/metabase/models/api_key.clj

```diff
@@ -0,0 +1,50 @@
+(ns metabase.models.api-key
+  (:require [crypto.random :as crypto-random]
+            [metabase.util.password :as u.password]
+            [methodical.core :as methodical]
+            [toucan2.core :as t2]))
+
+;; the prefix length, the length of `mb_1234`
+(def ^:private prefix-length 7)
+
+;; the total number of bytes of randomness we generate for API keys
+(def ^:private bytes-key-length 32)
+
+(methodical/defmethod t2/table-name :model/ApiKey [_model] :api_key)
+
+(doto :model/ApiKey
+  (derive :metabase/model)
+  (derive :hook/timestamped?))
+
+(defn prefix
+  "Given an API key, returns the standardized prefix for that API key."
+  [key]
+  (apply str (take prefix-length key)))
+
+(defn- add-prefix [{:keys [key] :as api-key}]
+  (cond-> api-key
+    key (assoc :key_prefix (prefix key))))
+
+(defn mask
+  "Given an API key, returns a string of the same length with all but the prefix masked with `*`s"
+  [key]
+  (->> (concat (prefix key) (repeat "*"))
+       (take (count key))
+       (apply str)))
+
+(defn generate-key
+  "Generates a new API key - a random base64 string prefixed with `mb_`"
+  []
+  (str "mb_" (crypto-random/base64 bytes-key-length)))
+
+(t2/define-before-insert :model/ApiKey
+  [{:keys [key] :as api-key}]
+  (-> api-key
+      add-prefix
+      (update :key u.password/hash-bcrypt)))
+
+(t2/define-before-update :model/ApiKey
+  [{:keys [key] :as api-key}]
+  (-> api-key
+      add-prefix
+      (update :key u.password/hash-bcrypt)))
```

Comment:
If the `key` stored in the DB is already a hashed key, why do we need to do that again on updates? Wouldn't that hash it again and it would become corrupted?

(I'm assuming that if the key name is updated or something, the FE would include the hashed value as the `:key` since it should never have access to the unhashed value after it's generated)

Reply:
I'm actually planning for the `regenerate-key` endpoint here. We'll want to update and set a new `key` on an existing API key, so I was imagining we would do something like `(t2/update! :model/ApiKey {:id ...} {:key (generate-a-new-key)})`.

I was not imagining that someone would do `(t2/update! :model/ApiKey {:id ...} an-api-key-selected-from-the-db)`. That seems undesirable to me, since you'll risk race conditions by possibly overwriting values that were set by other threads - much better to only explicitly update the particular fields you're interested in updating.

But if this is a concern, one option might be to stop overloading `:key` to mean "hashed key" *and* the unhashed key. You could only write to the `unhashed_key` field, and only read from the `hashed_key` field. Might be better anyway - WDYT?

Reply:
Oh that's fair. But yeah it's a somewhat common pattern in the code base to read a model, change a single field, and then write it back, even if you're not planning on doing that directly here. I like the idea of not overloading the `:key` field to make it less error-prone